### PR TITLE
Feature/oauth2 client registration

### LINKED ISSUE #36: [UserStory] As a developer, I can register a client application with whitelisted redirect URIs
### Story
As a developer, I want to register my client application with MyAuthService so that my app can participate in the OAuth2 flow.

### Context
Clients must be pre-registered with their allowed redirect URIs for security (RFC 6749 Section 2). The `client_type` determines authentication requirements: confidential clients have credentials, public clients do not. Registration is done via Django admin for MVP.

### Scope
- Define Client model with fields:
  - `client_id` (REQUIRED) - Unique public identifier, randomly generated.
  - `client_secret` (REQUIRED for confidential clients) - Securely generated credential.
  - `client_type` (REQUIRED) - `confidential` or `public` per RFC 6749 Section 2.1.
  - `name` (REQUIRED) - Human-readable application name (displayed on consent screen).
  - `redirect_uris` (REQUIRED) - List of allowed redirect URIs (at least one required).
  - `created_at` / `updated_at` - Timestamps for auditing.
- Allow client registration via Django admin.
- Generate `client_id` as a cryptographically secure random string (e.g., 32 characters).
- Generate `client_secret` as a cryptographically secure random string (e.g., 64 characters).
- Public clients should not have a `client_secret` (or have it set to null/empty).

### Acceptance Criteria
- [ ] Client model exists with all required fields.
- [ ] `client_type` field supports `confidential` and `public` values.
- [ ] Clients can be created, updated, and deleted via Django admin.
- [ ] `client_id` is unique and cryptographically random.
- [ ] `client_secret` is cryptographically random for confidential clients.
- [ ] `client_secret` is null/empty for public clients.
- [ ] Multiple redirect URIs can be registered per client (stored as list/array).
- [ ] At least one redirect URI is required per client.
- [ ] `created_at` and `updated_at` timestamps are auto-populated.

### Additional Notes

**Technical Notes:**
- Use `secrets.token_urlsafe(32)` or similar for generating `client_id`.
- Use `secrets.token_urlsafe(64)` or similar for generating `client_secret`.
- MVP: Client secret stored in plaintext. **Technical debt:** Hash secrets in Milestone 3.
- Consider using Django's `ArrayField` (PostgreSQL) or JSON field for `redirect_uris`.

### LINKED ISSUE #52: [Task] Create OAuth2 Client Model
### Background

- This task implements the core Client model for OAuth2 client application registration per RFC 6749 Section 2.
- The Client model serves as the foundation for the OAuth2 authorization flow.
- All client applications must be registered with specific attributes before they can participate in OAuth2 flows.
- RFC 6749 Section 2.1 defines two client types: confidential and public.

### Todo

- [ ] Create Client model in `oauth2/models.py` with all required fields
- [ ] Add `client_id` field:
  - `CharField(max_length=64, unique=True, db_index=True)`
  - Primary public identifier for the client
- [ ] Add `client_secret` field:
  - `CharField(max_length=128, blank=True, null=True)`
  - Only required for confidential clients
- [ ] Add `client_type` field:
  - `CharField(max_length=20, choices=[('confidential', 'Confidential'), ('public', 'Public')])`
  - Per RFC 6749 Section 2.1
- [ ] Add `name` field:
  - `CharField(max_length=255)`
  - Human-readable name displayed on consent screen
- [ ] Add `description` field (optional):
  - `TextField(blank=True)`
  - Optional description for admin reference
- [ ] Add `redirect_uris` field:
  - `JSONField(default=list)`
  - Stores list of allowed redirect URIs
- [ ] Add `is_active` field:
  - `BooleanField(default=True)`
  - Allows disabling clients without deletion
- [ ] Add timestamp fields:
  - `created_at = DateTimeField(auto_now_add=True)`
  - `updated_at = DateTimeField(auto_now=True)`
- [ ] Implement `__str__` method returning `f"{self.name} ({self.client_id[:8]}...)"`
- [ ] Implement `clean()` method to validate:
  - `redirect_uris` contains at least one URI
  - `client_secret` is null/empty for public clients
- [ ] Create and run Django migrations
- [ ] Write model tests to verify field constraints and validation

### Acceptance Criteria

- [ ] Client model exists in `oauth2/models.py`
- [ ] All required fields are present with correct types and constraints
- [ ] `client_id` field has unique constraint and database index
- [ ] `client_type` field has choices constraint ('public', 'confidential')
- [ ] `client_secret` is nullable (for public clients)
- [ ] `redirect_uris` stores a list and validates at least one entry
- [ ] Timestamps are auto-populated on create/update
- [ ] Model validation prevents invalid states (e.g., empty redirect_uris)
- [ ] Migrations apply successfully without errors
- [ ] `__str__` returns human-readable representation

### Technical Notes

- Use `models.JSONField` for `redirect_uris` (supported in Django 3.1+ for all databases including SQLite)
- `client_id` must be indexed for query performance during authorization
- For MVP, `client_secret` stored as plaintext; hashing deferred to Milestone 3
- Consider adding `Meta` class with:
  - `ordering = ['-created_at']`
  - `verbose_name = 'OAuth2 Client'`
  - `verbose_name_plural = 'OAuth2 Clients'`

### LINKED ISSUE #53: [Task] Implement Secure Client Credentials Generation
### Background

- Client credentials (`client_id` and `client_secret`) must be automatically generated when a client is created.
- Credentials must be cryptographically secure and unpredictable per security best practices.
- `client_id` must be globally unique across all registered clients.
- Per RFC 6749 Section 2.1:
  - **Confidential clients** are capable of maintaining credential confidentiality and MUST have a `client_secret`.
  - **Public clients** (e.g., SPAs, native apps) are incapable of maintaining credential confidentiality and MUST NOT have a `client_secret`.

### Todo

- [ ] Create a utility module `oauth2/utils.py` for credential generation functions
- [ ] Implement `generate_client_id()` function:
  - Use `secrets.token_urlsafe(32)` (produces ~43 characters)
  - Returns URL-safe base64-encoded string
- [ ] Implement `generate_client_secret()` function:
  - Use `secrets.token_urlsafe(48)` (produces ~64 characters)
  - Returns URL-safe base64-encoded string
- [ ] Implement uniqueness check for `client_id`:
  - Query database to check if `client_id` exists
  - Retry with new value if collision (max 5 attempts)
  - Raise exception if max retries exceeded
- [ ] Override Client model's `save()` method:
  - Generate `client_id` only if not already set (new instance)
  - Generate `client_secret` only for confidential clients on creation
  - Ensure credentials are never regenerated on updates
- [ ] Implement validation in `clean()` method:
  - Confidential clients MUST have `client_secret`
  - Public clients MUST NOT have `client_secret` (set to None if provided)
- [ ] Write unit tests:
  - Test `client_id` generation format and length
  - Test `client_secret` generation format and length
  - Test uniqueness constraint and retry logic
  - Test credentials not overwritten on update
  - Test public vs confidential client secret handling
- [ ] Document credential generation in code comments

### Acceptance Criteria

- [ ] `client_id` is automatically generated on client creation
- [ ] `client_id` is at least 32 characters, URL-safe
- [ ] `client_id` is unique across all clients (with retry on collision)
- [ ] `client_secret` is automatically generated for confidential clients only
- [ ] `client_secret` is at least 48 characters, URL-safe
- [ ] Public clients have `client_secret` set to `None` (enforced)
- [ ] Confidential clients MUST have `client_secret` (validated)
- [ ] Existing credentials are NOT overwritten on model updates
- [ ] Collision retry limit prevents infinite loops
- [ ] All generation logic is covered by unit tests

### Technical Notes

- `secrets` module uses OS-provided CSPRNG (cryptographically secure)
- URL-safe base64 encoding ensures credentials work in URLs and headers
- Minimum entropy: 32 bytes = 256 bits for `client_id`, 48 bytes = 384 bits for `client_secret`
- Never log or expose `client_secret` in error messages

### Related Issues

_No response_

### LINKED ISSUE #54: [Task] Configure Django Admin for Client Management
### Background

- For MVP, client registration and management will be done through Django admin interface.
- Admin interface must be user-friendly and display relevant client information.
- Security: `client_secret` should only be visible once at creation, then hidden.
- Admin should support creating, viewing, editing, and deleting clients.

### Todo

- [ ] Update `oauth2/admin.py` to register Client model
- [ ] Create custom `ClientAdmin` class extending `admin.ModelAdmin`
- [ ] Configure `list_display`:
  - `name`, `client_id_display` (truncated), `client_type`, `is_active`, `created_at`
- [ ] Configure `list_filter`:
  - `client_type`, `is_active`, `created_at`
- [ ] Configure `search_fields`:
  - `name`, `client_id`, `description`
- [ ] Configure `readonly_fields`:
  - `client_id`, `client_secret_display`, `created_at`, `updated_at`
- [ ] Configure `fieldsets` for organized form:
  - "Client Information": `name`, `description`, `client_type`, `is_active`
  - "Credentials": `client_id`, `client_secret_display`
  - "Redirect URIs": `redirect_uris`
  - "Timestamps": `created_at`, `updated_at`
- [ ] Implement `client_id_display` method (truncated for list view)
- [ ] Implement `client_secret_display` method:
  - Show "Not set" for public clients
  - Show "••••••••" (masked) for existing confidential clients
  - Show full secret only on first save (with copy warning)
- [ ] Configure `redirect_uris` widget:
  - Use `Textarea` with JSON formatting help text
- [ ] Add `save_model` override to handle one-time secret display
- [ ] Add custom admin actions if needed (e.g., "Deactivate selected clients")
- [ ] Test admin interface functionality:
  - Create confidential client
  - Create public client
  - Edit existing client
  - List/search/filter clients
- [ ] Add inline help text for fields

### Acceptance Criteria

- [ ] Client model appears in Django admin under "OAuth2" section
- [ ] Clients can be created via Django admin interface
- [ ] `client_id` is displayed but not editable after creation
- [ ] `client_secret` is masked after initial creation
- [ ] Public clients show "Not applicable" for client_secret
- [ ] List view shows: name, client_id (truncated), client_type, is_active, created_at
- [ ] Search works on name and client_id
- [ ] Filter works on client_type and is_active
- [ ] `redirect_uris` field accepts JSON list input
- [ ] Validation errors are displayed clearly
- [ ] Timestamps are read-only and auto-populated

### Technical Notes

- Use `help_text` on model fields for inline guidance
- Consider using `django-json-widget` for better JSON editing experience
- Add confirmation for destructive actions (delete, deactivate)
- Use `list_editable` sparingly (only `is_active` if at all)

### Related Issues

_No response_

### LINKED ISSUE #55: [Task] Implement Redirect URI Validation and Storage
### Background

- Each client must have one or more whitelisted redirect URIs for security per RFC 6749 Section 3.1.2.
- Redirect URIs are critical for preventing authorization code interception attacks.
- Validation occurs at two points:
  1. **Registration time**: Validate URI format and store.
  2. **Authorization time**: Validate incoming `redirect_uri` matches stored URI exactly.
- RFC 6749 Section 3.1.2 requirements:
  - MUST be an absolute URI
  - MUST NOT include a fragment component (`#`)
  - Comparison MUST be exact string match (RFC 6749 Section 3.1.2.2)
- RFC 8252 (OAuth 2.0 for Native Apps) allows:
  - `http://localhost` (loopback) for native apps
  - Private-use URI schemes (e.g., `com.example.app:/callback`)

### Todo

- [ ] Create validation module `oauth2/validators.py`
- [ ] Implement `validate_redirect_uri(uri)` function:
  - Validate URI is absolute (has scheme and host)
  - Validate no fragment component (`#`)
  - Validate scheme is `https`, `http` (localhost only), or private-use scheme
  - Return normalized URI or raise `ValidationError`
- [ ] Implement `validate_redirect_uris(uris)` function:
  - Validate list has at least one entry
  - Validate each URI using `validate_redirect_uri()`
  - Check for duplicates
  - Return validated list or raise `ValidationError`
- [ ] Add custom validator to Client model's `redirect_uris` field
- [ ] Implement `Client.is_valid_redirect_uri(uri)` method:
  - Perform exact string match against stored URIs
  - Return `True` if match found, `False` otherwise
- [ ] Implement `Client.get_redirect_uri(uri=None)` method:
  - If `uri` provided: validate and return if whitelisted
  - If `uri` not provided and only one registered: return that URI
  - If `uri` not provided and multiple registered: return `None` (error case)
- [ ] Update admin interface for redirect URI management:
  - Show validation errors inline
  - Provide format examples in help text
- [ ] Write comprehensive validation tests:
  - Valid HTTPS URIs
  - Valid HTTP localhost URIs
  - Invalid HTTP non-localhost URIs (production)
  - Invalid URIs with fragments
  - Invalid relative URIs
  - Duplicate URI detection
  - Exact match verification
  - Single vs. multiple URI resolution
- [ ] Document redirect URI requirements in README

### Acceptance Criteria

- [ ] Multiple redirect URIs can be registered per client
- [ ] At least one redirect URI is required (empty list rejected)
- [ ] URIs are validated on save:
  - Must be absolute URI (scheme + host)
  - Must NOT contain fragment (`#`)
  - Must use allowed scheme (https, http://localhost, private-use)
- [ ] Duplicate URIs are rejected
- [ ] `is_valid_redirect_uri()` performs exact string match
- [ ] `get_redirect_uri()` returns correct URI based on input
- [ ] Invalid URIs show clear error messages
- [ ] Admin interface displays validation errors

### Technical Notes

- **Exact match is critical**: Partial matching or prefix matching enables attacks
- **Fragment prohibition**: Fragments are not sent to server, enabling interception
- **HTTPS requirement**: Prevents token interception in transit (except localhost for development)
- **Multiple URIs**: Allow for different environments but each must be explicitly registered

### Related Issues

_No response_

### LINKED ISSUE #57: [Task] Write Comprehensive Client Model Tests
### Background

- The Client model is foundational to the OAuth2 implementation
- All model behaviors, validations, and edge cases must be tested
- Tests ensure RFC 6749 compliance is maintained through future changes

### Todo

- [ ] Create test file `tests/unit/test_client_model.py`
- [ ] Write tests for Client model fields:
  - All required fields are present
  - Field types and constraints are correct
  - Default values work as expected
- [ ] Write tests for credential generation:
  - `client_id` is auto-generated on creation
  - `client_id` is unique
  - `client_secret` is generated for confidential clients only
  - Public clients have null `client_secret`
  - Credentials not overwritten on update
- [ ] Write tests for redirect URI validation:
  - Valid HTTPS URIs accepted
  - Valid HTTP localhost URIs accepted
  - Invalid HTTP non-localhost rejected
  - Fragment URIs rejected
  - Relative URIs rejected
  - Empty list rejected
  - Duplicate URIs rejected
- [ ] Write tests for `is_valid_redirect_uri()`:
  - Exact match returns True
  - Non-match returns False
  - Case sensitivity verified
- [ ] Write tests for `get_redirect_uri()`:
  - With matching URI provided
  - With non-matching URI provided
  - With no URI and single registered
  - With no URI and multiple registered
- [ ] Write tests for model validation:
  - `clean()` method validation
  - `save()` method behavior
- [ ] Ensure tests run in CI pipeline

### Acceptance Criteria

- [ ] All model fields have corresponding tests
- [ ] Credential generation has 100% coverage
- [ ] Redirect URI validation has 100% coverage
- [ ] Edge cases are covered (empty values, invalid types, etc.)
- [ ] Tests pass in CI
- [ ] Tests are documented with clear descriptions

### Technical Notes

_No response_

### Related Issues

_No response_

### PR DESCRIPTION
**Pull request by gundambox**
### What
This PR implements the OAuth2 Client model and associated infrastructure, establishing the foundation for OAuth 2.0 authorization flows. The implementation includes comprehensive validation, security features, and administrative tooling for managing OAuth2 clients.

### Why
The Client model is a fundamental building block for OAuth 2.0 authorization flows. Before implementing authorization endpoints, we need a robust system to register and manage OAuth2 clients with proper credential handling, redirect URI validation, and security controls. This implementation follows RFC 6749 (OAuth 2.0) and RFC 8252 (OAuth 2.0 for Native Apps) specifications.

## Related Issue

Fixes #36 , #52 , #53 , #54 , #55 , #57

## Type of Change

- [x] Feature - New feature or enhancement

### Code Quality
- [x] Code follows the project's code style (PEP 8)
- [x] Code is formatted with `black`
- [x] Code passes `flake8` linting
- [x] No new warnings or errors introduced

### Testing
- [x] All tests pass (`make test`)
- [x] New tests added for new features or bug fixes
- [x] Existing tests updated if needed

### Documentation
- [x] Documentation updated if needed (README, docstrings, etc.)
- [x] Comments added for complex logic
- [x] CONTRIBUTING.md updated if workflow changed

### Git
- [x] Commits follow the commit message format
- [x] Branch is up-to-date with `main`
- [x] Branch name follows naming conventions

## Testing Steps

1. Start the development environment:
   ```bash
   make build
   make up
   make migrate
   ```

2. Run the complete test suite:
   ```bash
   make test
   ```

3. Test admin interface functionality:
   ```bash
   make shell
   python manage.py createsuperuser
   ```
   - Navigate to `http://localhost:8000/admin/oauth2/client/`
   - Create a confidential client and verify one-time secret display
   - Create a public client and verify no secret is generated
   - Test redirect URI validation with valid and invalid URIs
   - Test bulk activate/deactivate actions

4. Verify redirect URI validation:
   - Attempt to create client with invalid HTTP URI (non-localhost)
   - Attempt to create client with fragment in URI
   - Verify private URI schemes are accepted
   - Verify duplicate URIs are rejected

### CODE REVIEW

**Comment by gundambox** [comment]
@codex review.

File: oauth2/validators.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            raise ValidationError(f"Redirect URI must be absolute: {uri}")
+
+    if parsed.fragment:
+        raise ValidationError(f"Redirect URI must not contain a fragment: {uri}")
+
+    scheme = parsed.scheme.lower()
+    netloc = parsed.netloc.lower()
+
+    if scheme == "https":
+        return uri
```
[the remark this answers is not in the document]
Comment (gundambox) [reply oauth2/validators.py:27]: Solve at 0293be3ed4934440d0c20e61a63b87d6c8eec7bd

File: oauth2/admin.py [hunk 2]
```diff
@@ -1 +1,151 @@
...
+            request.session[f"new_client_secret_{obj.pk}"] = obj.client_secret
+            request.session.modified = True
+
+    def changeform_view(self, request, object_id=None, form_url="", extra_context=None):
+        # Check if we have a newly created secret to display
+        if object_id:
+            session_key = f"new_client_secret_{object_id}"
+            if session_key in request.session:
+                self._display_secret = request.session.pop(session_key)
+                request.session.modified = True
```
[the remark this answers is not in the document]
Comment (gundambox) [reply oauth2/admin.py:111]: use extra_context to refactor this section

**Comment by gundambox** [comment]
@codex review.

**Comment by gundambox** [comment]
@codex review.

File: oauth2/validators.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        if hostname in ["localhost", "127.0.0.1", "::1"]:
+            return uri
+        else:
+            raise ValidationError(
+                f"HTTP redirect URIs are only allowed for localhost: {uri}"
+            )
+    else:
+        # Allow private URI schemes (RFC 8252)
+        # e.g., com.example.app:/callback, myapp://callback
+        return uri
```
[the remark this answers is not in the document]
Comment (gundambox) [reply oauth2/validators.py:45]: This project is currently focused on implementing RFC 6749 (OAuth 2.0 Authorization Framework) and RFC 6750 (Bearer Token Usage). Support for RFC 8252 (OAuth 2.0 for Native Apps) is out of scope at this time and will be considered for future iterations.
